### ISSUE TITLE
Send NAT hole punch packets to trigger forwarding of RTP at some Relays & SBCs

### ISSUE THREAD

**Issue by mcguinn** [jitsi/jigasi#30]
This is a feature request. 

As soon as the Jigasi SIP client state is connected Jigasi should send a few UDP NAT hole punch packets. This will help unlock SBCs or Relays that do not forward RTP until they see RTP from both legs. And once the SBCs start forwarding media then the conference made up of only SIP clients connected to the same SBC/Relay will start mixing and sending conferences streams to the endpoints. Ideally this behaviour should be configurable (enable/disable, number of hole punch packets)

I encountered this scenario in my testing behind and quickly hacked a workaround but would prefer to see this in an official Jigasi build

Here is the rough outline of my hack (comments tagged "TWH"). It addresses only audio but  real solution should address both audio and video:

			- in GatewaySession.peerStateChanged
				if (CallPeerState.CONNECTED.equals(callPeerState))
				{
					// TWH - try to stimulate media at SBC
					Util.sendHolePunchPacket(evt.getSourceCallPeer().getCall());
				}
	
			- in Jitsi CallPeerMediaHandler
			   public void sendHolePunchPacket(MediaStreamTarget target)
				{
					getTransportManager().sendHolePunchPacket(target, MediaType.AUDIO); // TWH was VIDEO
				}
				
			- in Jitsi TransportManager

			public void sendHolePunchPacket(MediaStreamTarget target, MediaType type)
			{
				logger.info("Send NAT hole punch packets");

				//check how many hole punch packets we would be supposed to send:
				int packetCount
					= ProtocolMediaActivator.getConfigurationService().getInt(
							HOLE_PUNCH_PKT_COUNT_PROPERTY,
							DEFAULT_HOLE_PUNCH_PKT_COUNT);

				if (packetCount < 0)
					packetCount = DEFAULT_HOLE_PUNCH_PKT_COUNT;
				if (packetCount == 0)
					return;

				try
				{
					StreamConnector connector = getStreamConnector(type);

					if(connector.getProtocol() == StreamConnector.Protocol.TCP)
						return;
					// TWH - use a real RTP packet
					byte[] buf = new byte[] {
							(byte)0x80, 0x7C,
							0x7B, 0x77,
							0x35, (byte)0xAA, (byte)0xF1, 0x1C,
							0x37, (byte)0xDE, 0x00, 0x36,
							0x50, 0x15, (byte)0xC7, 0x70, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x10, 0x11, 0x12,
							0x50, 0x15, (byte)0xC7, 0x70, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x10, 0x11, 0x12,
							0x50, 0x15, (byte)0xC7, 0x70, 0x01 };

					synchronized(connector)
					{
						//we may want to send more than one packet in case they get lost
						for(int i=0; i < packetCount; i++)
						{
							DatagramSocket socket;

							// data/RTP
							if((socket = connector.getDataSocket()) != null)
							{
								InetSocketAddress dataAddress = target.getDataAddress();

								socket.send(
										new DatagramPacket(
												buf,
												buf.length,
												dataAddress.getAddress(),
												dataAddress.getPort()));
							}

							// control/RTCP
							if((socket = connector.getControlSocket()) != null)
							{
								InetSocketAddress controlAddress
									= target.getControlAddress();

								socket.send(
										new DatagramPacket(
												buf,
												buf.length,
												controlAddress.getAddress(),
												controlAddress.getPort()));
							}
						}
					}
				}
				catch(Exception e)
				{
					logger.error("Error cannot send to remote peer", e);
				}
			}

**Comment by emcho** [comment]
We should already have this as we inherit it from Jitsi. Maybe we disabled
it for some reason ...

Just to make sure we are on the same page: you should only need this in
cases where none of the WebRTC clients are sending audio. If the are, then
their audio packets would effectively punch the holes.

On Monday, 18 April 2016, mcguinn <notifications@github.com> wrote:

> This is a feature request.
>
> As soon as the Jigasi SIP client state is connected Jigasi should send a
> few UDP NAT hole punch packets. This will help unlock SBCs or Relays that
> do not forward RTP until they see RTP from both legs. And once the SBCs
> start forwarding media then the conference made up of only SIP clients
> connected to the same SBC/Relay will start mixing and sending conferences
> streams to the endpoints. Ideally this behaviour should be configurable
> (enable/disable, number of hole punch packets)
>
> I encountered this scenario in my testing behind and quickly hacked a
> workaround but would prefer to see this in an official Jigasi build
>
> Here is the rough outline of my hack (comments tagged "TWH"). It addresses
> only audio but real solution should address both audio and video:
>
>         - in GatewaySession.peerStateChanged
>             if (CallPeerState.CONNECTED.equals(callPeerState))
>             {
>                 // TWH - try to stimulate media at SBC
>                 Util.sendHolePunchPacket(evt.getSourceCallPeer().getCall());
>             }
>
>         - in Jitsi CallPeerMediaHandler
>            public void sendHolePunchPacket(MediaStreamTarget target)
>             {
>                 getTransportManager().sendHolePunchPacket(target, MediaType.AUDIO); // TWH was VIDEO
>             }
>
>         - in Jitsi TransportManager
>
>         public void sendHolePunchPacket(MediaStreamTarget target, MediaType type)
>         {
>             logger.info("Send NAT hole punch packets");
>
>             //check how many hole punch packets we would be supposed to send:
>             int packetCount
>                 = ProtocolMediaActivator.getConfigurationService().getInt(
>                         HOLE_PUNCH_PKT_COUNT_PROPERTY,
>                         DEFAULT_HOLE_PUNCH_PKT_COUNT);
>
>             if (packetCount < 0)
>                 packetCount = DEFAULT_HOLE_PUNCH_PKT_COUNT;
>             if (packetCount == 0)
>                 return;
>
>             try
>             {
>                 StreamConnector connector = getStreamConnector(type);
>
>                 if(connector.getProtocol() == StreamConnector.Protocol.TCP)
>                     return;
>                 // TWH - use a real RTP packet
>                 byte[] buf = new byte[] {
>                         (byte)0x80, 0x7C,
>                         0x7B, 0x77,
>                         0x35, (byte)0xAA, (byte)0xF1, 0x1C,
>                         0x37, (byte)0xDE, 0x00, 0x36,
>                         0x50, 0x15, (byte)0xC7, 0x70, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x10, 0x11, 0x12,
>                         0x50, 0x15, (byte)0xC7, 0x70, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x10, 0x11, 0x12,
>                         0x50, 0x15, (byte)0xC7, 0x70, 0x01 };
>
>                 synchronized(connector)
>                 {
>                     //we may want to send more than one packet in case they get lost
>                     for(int i=0; i < packetCount; i++)
>                     {
>                         DatagramSocket socket;
>
>                         // data/RTP
>                         if((socket = connector.getDataSocket()) != null)
>                         {
>                             InetSocketAddress dataAddress = target.getDataAddress();
>
>                             socket.send(
>                                     new DatagramPacket(
>                                             buf,
>                                             buf.length,
>                                             dataAddress.getAddress(),
>                                             dataAddress.getPort()));
>                         }
>
>                         // control/RTCP
>                         if((socket = connector.getControlSocket()) != null)
>                         {
>                             InetSocketAddress controlAddress
>                                 = target.getControlAddress();
>
>                             socket.send(
>                                     new DatagramPacket(
>                                             buf,
>                                             buf.length,
>                                             controlAddress.getAddress(),
>                                             controlAddress.getPort()));
>                         }
>                     }
>                 }
>             }
>             catch(Exception e)
>             {
>                 logger.error("Error cannot send to remote peer", e);
>             }
>         }
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/jitsi/jigasi/issues/30>
>


-- 
sent from my mobile

**Comment by mcguinn** [comment]
Yes. Correct. WebRTC clients use ICE to negotiate direct P2P media and when they are involved in the call then there is no need for the hole punching as the conference has media to forward and this unlocks our SBC for any SIP clients that are involved in the conference. But when its a conference exclusively composed of SIP clients whose media is relayed by our SBC then this is the case where we need this hole punching. 

I forget where I saw this code fragment used I think I  tried to enable is-as but it had no effect until I added in the other aspects of the hack. After this media started passing through the SBC and the conference worked.

**Comment by emcho** [comment]
It doesn't really matter if there are sip only agents. What matters is whether some of them would be recvonly. Now that I am thinking about this: jigasi shouldn't need this as it has its own mixer which should be generating packets even if none of the endpoints behind it is sending anything.

**Comment by mcguinn** [comment]
I also would expect the behaviour that you describe. Jigasi should send comfort noise or something similar towards the the conference bridge but that does not appear to be the case. Nothing is coming out of Jigasi until we punch holes and media from the endpoints starts arriving at Jigasi/conference bridge. Unfortunately I did not dig deeply enough to understand the root cause of this problem but just assumed that without incoming media the conference bridge has nothing to mix and consequently Jigasi has nothing to send to the endpoints.
